### PR TITLE
Update to C++14.

### PR DESCRIPTION
**Pull request by jonsimantov**
The last release added a warning when building with older versions of C++ such as C++11. This update (which will be released with the next major version) changes the C++ build to use the C++14 standard, and removes the warning added in the previous release.
***

Integration tests - a few known flakes / errors. Packaging has also been run and succeeded.
***

### Type of Change
Place an `x` the applicable box:
- [ ] Bug fix. Add the issue # below if applicable.
- [ ] New feature. A non-breaking change which adds functionality.
- [x] Other, such as a build process or documentation change.
***

#### Notes
- Bug fixes and feature changes require an update to the `Release Notes` section of `release_build_files/readme.md`.
- Read the contribution guidelines [CONTRIBUTING.md](https://github.com/firebase/firebase-cpp-sdk/blob/main/CONTRIBUTING.md).
- Changes to the public API require an internal API review. If you'd like to help us make Firebase APIs better, please propose your change in a feature request so that we can discuss it together.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 cmake_minimum_required (VERSION 3.1)
 
-set (CMAKE_CXX_STANDARD 11)
+set (CMAKE_CXX_STANDARD 14)
```
[the remark this answers is not in the document]
Comment (jonsimantov) [reply CMakeLists.txt:19]: Good call.

**Comment by jonsimantov** [comment]
Firestore Android build is broken by this update: [log here - give it a moment to load, the log file is large](https://github.com/firebase/firebase-cpp-sdk/actions/runs/4569195737/jobs/8065164371#step:8:66419)

**Comment by dconeybe** [comment]
> Firestore Android build is broken by this update

Drive-by comment: Try changing `make_unique` to `absl::make_unique`. C++14 adds `std::make_unique` and both must be in scope, resulting in the ambiguous reference. In C++11, `std::make_unique` didn't exist.

**Comment by dconeybe** [comment]
Here is the make_unique that appears to need fixing: https://github.com/firebase/firebase-cpp-sdk/blob/73ce6feb70d3e830676aafa1d0ded64a57f07fb8/firestore/integration_test_internal/src/android/promise_android_test.cc#L213

**Comment by wu-hui** [comment]
This PR won't work with Firestore yet.

A working PR already exists: https://github.com/firebase/firebase-cpp-sdk/pull/1182

@cherylEnkidu You should merge that one first before this PR can work.

**Comment by jonsimantov** [comment]
> This PR won't work with Firestore yet.
> 
> A working PR already exists: #1182
> 
> @cherylEnkidu You should merge that one first before this PR can work.

We're not merging this into main yet - that PR should be merged into this one.

**Review by a-maurice — APPROVED** [review]

**Comment by a-maurice** [comment]
I assume this is waiting for the release that will include it.  Also, we probably do want to update the release notes for this. Though I could see that in a different CL that updates the docs and such.

**Comment by jonsimantov** [comment]
> I assume this is waiting for the release that will include it. Also, we probably do want to update the release notes for this. Though I could see that in a different CL that updates the docs and such.

Added an update to the readme as well.

**Comment by a-maurice** [comment]
Do you want to remove the header warning as well?

**Comment by jonsimantov** [comment]
> Do you want to remove the header warning as well?

Yeah, good call.

**Pull request merged**